text/0000-embroider-v2-package-format.md

```diff
@@ -0,0 +1,817 @@
+
+Another example is that a package like `fastboot` could expose config that says whether you're in Fastboot vs the browser, so that apps and addons could say `getConfig('fastboot').isFastboot`. Notice that for an addon to do this, it should declare `fastboot` as an optional peerDependency (see **Optional Peer Dependencies** section), because `getConfig` can only see your **Allowed Dependencies**.
+
+### JavaScript Macro: macroCondition
+
+`macroCondition` acts like a function that takes a boolean value and returns that same boolean value. But whenever `macroCondition` appears directly inside the predicate of an `if` statement or as the predicate of a ternary expression, it tells the macro system to do branch elimination based on the predicate. Here is an example that combines three of the macros we've seen so far:
+
+```js
+// This example:
+import { macroCondition, getOwnConfig, importSync } from "@ember/macros";
+
+let implementation;
+
+if (macroCondition(getOwnConfig().useNewVersion)) {
+  implementation = importSync("./new-component");
+} else {
+  implementation = importSync("./old-component");
+}
+
+export default implementation;
+
+// ==============
+// Could compile down to this if OwnConfig contains { useNewVersion: true }
+let implementation;
+implementation = importSync("./new-component");
+export default implementation;
+
+// ===============
+// Or compile down to this if OwnConfig contains { useNewVersion: false }
+let implementation;
+implementation = importSync("./old-component");
+export default implementation;
+```
+
+It is a build error if `macroCondition` cannot statically determine the truth status of its argument. It's a build error if `macroCondition` appears anywhere other than as the predicate of an if statement or ternary expression.
+
+`macroCondition` supports boolean logic, like `macroCondition(getOwnConfig().a && getOwnConfig().b)`.
+
+Here is an example of `macroCondition` in a ternary expression:
+
+```js
+// this example:
+const flavor = macroCondition(getOwnConfig().prefersChocolate)
+  ? "chocolate"
+  : "vanilla";
+// could compile down to:
+const flavor = "chocolate";
+```
+
+`macroCondition` is the foundation that lets us choose which code to include in the build. You can choose to inline two different implementations within the branches of an `if` statement, or you can keep them in entirely separate modules and import only the correct one via `importSync`.
+
+It would also be possible (in the future, when top-level await stabilizes) to use [top-level await](https://github.com/tc39/proposal-top-level-await) to replace usage of our `importSync` macro with standards-compliant `import()`:
+
+```js
+if (macroConditional(getOwnConfig().x)) {
+  await import("x");
+} else {
+  await import("y");
+}
+```
+
+Q: Why not allow `if (getOwnConfig().thing)` instead of `if (macroCondition(getOwnConfig().thing))`?
```

Comment:
Isn't this still "allowed"? `macroCondition` is just an optional "lint" isn't it?

Reply:
Correct, it would still be allowed and it *might* even get your branch elimination if a tool like uglify sees the output of this.

It's a little more than a lint because it does change the point within the build process at which branch elimination happens. With macroCondition, branch elimination will happen *before* you traverse beyond the current module to discover more modules. So if you eliminate a branch that imports a non-parseable file, macroCondition prevents the parse error whereas relying on uglification would not.